llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2271,6 +2271,27 @@ Instruction *InstCombinerImpl::foldVectorBinop(BinaryOperator &Inst) {
     }
   }
 
+  // Similar to the combine above, but handles the case for scalable vectors
+  // where both V1 and C are splats.
+  //
+  // Op(shuffle(V1, 0), (splat C)) -> shuffle(Op(V1, (splat C)), 0)
```

Comment:
This transformation may introduce immediate UB.
Counterexample: `sdiv (splat C), shuffle(V1, 0) -> shuffle(Op((splat C), V1), 0)` is invalid if `V1` has a zero value at a non-zero index.

Reply:
There's a check for `isSafeToSpeculativelyExecuteWithVariableReplaced` earlier on in the function which the previous combine also relies on, so I think this should also be covered in the new combine. I've added a negative test in e9d51e74cfeb274d263d85148e74b31937ca49dd

Reply:
Right, but the transform you're modeling off just above has several other checks for introducing UB.  The mask logic is one, the getSafeVectorConstantForBinop is another.  I'm not sure the mask reasoning is needed for this one, but the divrem/shift definitely is?  If not, why is the previous block required?

Reply:
getSafeVectorConstantForBinop is needed in the previous block because if the mask has undef then the "new constant" will have undef for those lanes, which need swapped out e.g:

```llvm
define <4 x i32> @sdiv_sdiv_undef_mask_elt(<4 x i32> %v0) {
; CHECK-LABEL: @sdiv_sdiv_undef_mask_elt(
; CHECK-NEXT:    [[T3:%.*]] = sdiv <4 x i32> [[V0:%.*]], <i32 1, i32 2, i32 7, i32 1>
; CHECK-NEXT:    ret <4 x i32> [[T3]]
;
  %t1 = sdiv <4 x i32> %v0, <i32 1, i32 2, i32 3, i32 4>
  %t2 = sdiv <4 x i32> %v0, <i32 5, i32 6, i32 7, i32 8>
  %t3 = shufflevector <4 x i32> %t1, <4 x i32> %t2, <4 x i32> <i32 undef, i32 1, i32 6, i32 undef>
  ret <4 x i32> %t3
}
```

We don't have that problem in the new combine because the mask is always zero